### ISSUE TITLE
Potential LLM Bias

### ISSUE THREAD

**Issue by alexander-hanel** [Noelo-Lab/decbench#43]
> DecBench v1.0 — decompiler benchmarking suite · projects: base-passwd, bash, betaflight, bzip2, chibios, cleanflight, coreutils, crazyflie, cronie, dexter, diffutils, dpkg, e2fsprogs, findutils, freertos, gnutls, grep, gzip, iproute2, kmod, libacl, libbsd, libedit, libexpat, libopencm3, libselinux, minipig, mirai, mydoom, nuttx, openssh-portable, riot-os, rsyslog, shadow, sysvinit, tar, u-boot, x0r-usb, zlib

Hello, is it possible that the benchmarking suite is skewed towards LLMs because the source code was included in their training data?

**Comment by mahaloz** [comment]
@alexander-hanel, it's really quite hard to quantify. I would 100% agree that every project in this list is in the AI training set. The assumption here is that if they realize they are working on these projects, they could just recall the code.

So, there are some mitigations here:
1. Remove all symbols (stripping)
2. Anonymize binary names
3. Optimize: it _significantly_ changes the assembly and associated decompilation

That third case is where many interesting things happen. Take a look at a few of the `02` (both inline and no inline) for Codex:
https://decbench.com/view/?tier=sample-set&dec=codex&metric=ged&fn=sysvinit%2FO2-noinline%2Futmpdump%3A%3Adump

In many cases, it just straight up fails.
In others, you will notice that it does not recognize the original source, and instead tries to make code that is actually _better_ (to me) than it, causing it to actually lose on GED. 

I think the future of evaluating how much they are cheating will be very very hard. I plan to do a closed dataset, but, even in that case, it's impossible to know what could have been trained. If I construct my own source/binaries, than we also make unrealistic programs. We are in quite the pickle for future evaluations here, so this is a best attempt :).

**Comment by alexander-hanel** [comment]
@mahaloz, I agree that the first two mitigations are necessary, especially because symbol names, binary names, and embedded strings could provide clues about the binary’s origin. Once tokenized, these identifiers may allow an LLM to associate the binary with source code or projects encountered during training. The optimize mitigation creates another level of complexity that I doubt LLMs or even many decompilers handle  well. 

For evaluation/post-analysis, I use a final prompt to create an [after-actions review](https://en.wikipedia.org/wiki/After-action_review) to identify an fallacies, issues, or opportunities for improvements. Within yours, it might be useful to ask the LLM if it was able to identify the source of the binary during the decompilation process. It could provide some clues if the LLM was able to recall the source. 

Cheers. 

Feel free to close out this issue.

**Comment by henrycharlesworth** [comment]
Basically I don't think there's anything you can do about this issue since they are just trained on essentially everything that is open source. But just to give an example:

<img width="2201" height="863" alt="Image" src="https://github.com/user-attachments/assets/6de30b32-090a-40e2-852e-405d3fd688dc" />

Claude has generated completely identical code, including the comments and all variable names. So very hard to quantify but yeah it will have an effect

**Comment by pr0me** [comment]
From our own experiments, I assume that part of the reason why the LLMs are so good at matching functions to their groundtruth might be that they have full access to the whole binary during their agentic workflow (if I understand the methodology correctly).

One way to address this would be to supply only the raw disassembly of the isolated target function, not the whole binary. This of course has its own caveats, as, e.g., information for full type recovery might be lacking within a limited view, but it might provide interesting insights on its own. Maybe not replace the current agent methodology but introduce this as a separate category.

**Comment by mahaloz** [comment]
I've reopened this issue because I've found it interesting :). 

### Investigation

@henrycharlesworth I've done some investigation on the server side. Here is some text directly from the Claude Code trace:
```
Done. The target function at `0xf510` is bzip2's **`fallbackSort`**.
```

Without instructions, it just went ahead and guessed whose code this was. I find this a little wild, considering it only used 3 tool calls. All to objdump. No internet lookup. No HuggingFace-breakout moment or anything. It just recalled based on RAW ASSEMBLY. I find this DOUBLE wild because it got the version correct too. This function changes with version. 

This is annoying because I looked through the Codex logs on the same function, and it never concludes it knows this binary. All of this means detecting _cheating_ (which I had defined as using a web browser or finding the source) is even harder than I previously thought it would be.

To make this more transparent, I've updated the public results with the traces. Find them here:
https://huggingface.co/datasets/noelo-lab/decbench-dataset/tree/main/results/claude-code/traces

### Mitigations

I have a few ideas, but they're all band-aids. I think, regardless, I'll add a little warning to the page that frontier LLM results can be biased even with mitigations.

Ideas:
1. **Encrypt all of the strings.** I have not quantified it yet, but so far, the majority of recall appears to happen with an LLM finding strings it knows in the binary (some that don't even say the binary name).
2. **Ask the LLM if it knows the target.** If it returns the exact project, it gets marked for cheating. Unsure how to represent this on the leaderboard yet. This is not a failure, but this is also not a win. We may just have to give them a small denominator of binaries, which scews accuracy. 
3. **No agents, only LLM.** Use @pr0me idea: stop it from doing agentic lookup of other functions. Likely stops them from finding info, but also diminishes the real reason they are good now. Hard choice. 

Open to other ideas.

**Comment by henrycharlesworth** [comment]
I mean you can try to ask it not to do things from recall - like maybe explicitly build this into the rules you give it in the prompt. But this is only going to help so much

**Comment by delicious-pancakes** [comment]
> I mean you can try to ask it not to do things from recall - like maybe explicitly build this into the rules you give it in the prompt. But this is only going to help so much

There's no way to prevent an LLM from using its own knowledge, it will just agree with you and continue to do it, despite already being polluted, the best you can do is prevent it from acquiring knowledge its not supposed to have, and even then, unless you have a sterile sandbox enviroment, it will find ways to circumvent that too

What you likely need is separate leaderboards for LLM free disasm and LLM assisted disasm and just assume that the LLMs are going to cheat and use code with no publically available source of truth for scoring

**Comment by stevenagy** [comment]
We've also faced this, whether on whole-binary disassembly or one function at a time (stripped). Joining here as I'm curious what ideas other folks have tried. While we're not doing decompiler testing, I think one workaround for you might be to use synthetically generated programs.

**Comment by mahaloz** [comment]
@stevenagy interesting... 

I'm working on some follow-up to this problem after observing the issues. I'll update here early if I can get it to arxiv in a timely manner. 

There are some interesting things here.

**Comment by jeffmcjunkin** [comment]
Maybe make your own binaries, that aren't part of the model's training data, that you then do the evaluations upon? Not quite just "Hey Claude, migrate `passwd` to Rust. Make no mistakes." but generally making moderately-sized projects that you do the evaluations upon?

**Comment by mahaloz** [comment]
@jeffmcjunkin, the issue is that true evaluation of decompilers comes from the program complexity found in _real_ programs, with variety across types. It's hard to mimic in practice, and multiple research areas have tried and failed to do it properly. I doubt I alone will be able to. 

One alternative idea is to mutate the open-source code to preserve the complexity but make it different enough to throw off recall. All of the above present novel areas that are hard to do right.

**Comment by alexander-hanel** [comment]
> Joining here as I'm curious what ideas other folks have tried

@stevenagy, an approach I’m using to measure how well models handle complexity in reverse engineering is generate binaries in which I can incrementally increase the complexity of the code to see how well it can write decryptors. So far I have only tested on local models but it’s been a fun project. Here is a link to the results. 

https://alexander-hanel.github.io/StressingLLMs/

**Comment by sjkim1127** [comment]
This is just a hypothesis that came up while I was analyzing this issue with the help of AI.

The example above where Claude reproduced the original variable names made me wonder if source-only identifier recovery might be useful to track separately.

If an unusual local variable name is verified to be absent from the stripped binary — including debug info, symbols, .rodata, assert/log strings, etc. — but an LLM repeatedly reconstructs the exact original identifier, that could be a useful signal of source familiarity.

This obviously would not prove memorization, since some names may be inferred from semantics or algorithm recognition. I also wouldn’t treat it as cheating or use it to penalize the score. It might simply be useful as an additional diagnostic for cases where source-specific prior knowledge may be contributing to the reconstruction.

**Comment by sjkim1127** [comment]
Rather than generating non-idiomatic or inefficient code synthetically, another approach would be to mine real historical implementations with verifiable human provenance and, where possible, a reasonably bounded history of public availability.

A particularly interesting case would be a short-lived, inefficient or non-idiomatic implementation that was later replaced by a cleaner equivalent which then remained in use for many years. I would probably favor low-profile cleanup or refactoring changes over security fixes or widely discussed bugs, since the earlier versions may otherwise have received substantial secondary exposure.

An especially useful control might be older human-written code that was only made public after a model’s declared knowledge cutoff, although such examples would probably be rare.

This could avoid some of the “unrealistic program” problem while testing whether an LLM decompiler reconstructs the actual historical implementation, or instead gravitates toward a more canonical implementation despite that form postdating the model’s declared knowledge cutoff.
